### DISCUSSION TITLE
Running pesq on batches of audio

### DISCUSSION THREAD

**Discussion by giamic** [ludlows/python-pesq#38] [Ideas]
Hi all,
I'm working with batches of audio files stored in 2D numpy arrays with shape `(batch_size, num_samples)`. Currently, the only way I found to run pesq through the batch is with a for loop but it is quite slow. Would it be possible to add a vectorized version of pesq that takes batches of audio in input and returns an array of pesq scores in output?

**Comment by ludlows** [comment]
thanks for your idea.
it is totally doable. 
however, what behavior is supposed to be introduced once an error occurs in the batch processing?

**Reply by giamic** [reply]
Thank you for the quick reply!
That's a very good question. I can see two alternatives:
- Raise a `PesqError` when any of the audios in the batch returns an error. The advantage is that it is very simple to implement but in case the error is raised all the information contained in the batch will probably be lost.
- Return a special value, for example one among `[None, np.nan, -np.inf, -1, 0, 1]`. Each of these possibilities has their own pros and cons. But in general, this approach might be too silent a way to deal with errors. Catching the error message and logging it out as a warning might be an acceptable compromise?

I think that raising the error on the entire batch is acceptable. Among the "special values", my favourite one would be `None` I guess, but I don't have a super strong opinion.

**Reply by ludlows** [reply]
Thanks for your answer.
I prefer the behavior to raise an error as well.
But let us keep the discussion open for a while and listen to other opinions.

**Reply by giamic** [reply]
Hi, I see that this discussion hasn't gathered much attention unfortunately. Are you still open to implement this feature however?

**Reply by ludlows** [reply]
> Hi, I see that this discussion hasn't gathered much attention unfortunately. Are you still open to implement this feature however?

yes. I am planing to implement the batch version on the dev branch.

**Reply by ludlows** [reply]
now, on the `dev` branch, we can import `pesq_batch` from `pesq` module.  
 
please check out the `dev` branch and have a try.

**Reply by giamic** [reply]
Thank you for your work! I've tested it and it seems to work as intended. This is a nice convenience function to have. However, it's not faster than simply looping over the different elements of the batch. Actually, I took a look at the source code and it seems this is exactly what it's doing.

My initial request was more something that would probably mean modifying the C code behind to provide benefits similar to the ones we have when using numpy for array operations instead of looping in pure python.

**Reply by ludlows** [reply]
thanks for your reply.
it seems the vectorized version of pesq still needs to copy audio values in a `for` loop, which is hard to boost the performance (time efficience) in a large step.  It is possible to implement the `for` loop inside the `cython` part. But I am not sure that is what you want.

**Reply by giamic** [reply]
To be honest with you, I don't know how to solve this (I would have made a PR myself otherwise :D ). I just suspect that it is possible to have a much faster implementation because the calculation on different elements of the batch can be parallelised, a bit like numpy does. I would expect a strongly sublinear increase of computation time when adding new elements of the batch.

This sublinear performance is extremely interesting when doing for example statistical analysis of the results of a speech enhancement algorithm, a situation in which we might have to calculate pesq on thousands of files

**Reply by ludlows** [reply]
as you described, the proper solution is to utlize the `mutiprocessing` feature in python.

**Reply by ludlows** [reply]
hi @giamic ,
the `multiprocessing` feature is available now on `dev` branch.
you could try it with thousands of files if you are interested.
thanks.

**Reply by giamic** [reply]
This is great! On my local machine it took 4 times less to compute using the new algorithm.
I tested with the following code:
```python
if __name__ == "__main__":
    fs = 16000
    duration = 10
    batch_size = 64
    input_shape = (batch_size, fs * duration)
    x = np.random.random(input_shape)
    y = np.random.random(input_shape)
    n_processors = multiprocessing.cpu_count()
    print(f"Available processors: {n_processors}")
    t0 = time.time()
    res = []
    for i in range(batch_size):
        res.append(pesq(fs, x[i], y[i], "wb"))
    print(f"Time without multithreading: {time.time() - t0:.3f} s")
    t1 = time.time()
    res_batch = pesq_batch(fs, x, y, "wb", n_processor=n_processors)
    print(f"Time with multithreading   : {time.time() - t1:.3f} s")
    assert np.allclose(res_batch, res)
```
and the output was:
```
Available processors: 8
Time without multithreading: 15.320 s
Time with multithreading   : 4.344 s

Process finished with exit code 0
```

This is brilliant. Thank you!

**Reply by giamic** [reply]
The only suggestion I would make is to use `multiprocessing.cpu_count()` as the default value for `n_processor` instead of None, but I guess that this is a matter of taste.

**Reply by ludlows** [reply]
thanks for your suggestion and the work on testing this feature.

about the default value of  `n_processor`,  I agree with your suggestion.
